CM7/Core/Src/ssi_encoder.c

```diff
@@ -0,0 +1,57 @@
+#include "ssi_encoder.h"
+
+#include <stdlib.h>
+#include <assert.h>
+#include <math.h>
+
+#include "stm32h7xx.h"
+#include "stm32h7xx_hal_spi.h"
+
+#define ENCODER_MAX_COUNT ((1 << 13) - 1)
+
+ssi_encoder_t *ssi_encoder_init(SPI_HandleTypeDef *spi_handle)
+{
+    // make sure the handle exists
+    if (spi_handle == NULL)
+        return NULL;
+
+    // set required configuration for ssi encoder
+    spi_handle->Init.Mode = SPI_MODE_MASTER;
+    spi_handle->Init.Direction = SPI_DIRECTION_2LINES_RXONLY;
+    spi_handle->Init.DataSize = SPI_DATASIZE_13BIT;
+    spi_handle->Init.CLKPolarity = SPI_POLARITY_LOW;
+    spi_handle->Init.CLKPhase = SPI_PHASE_1EDGE;
+    spi_handle->Init.NSS = SPI_NSS_SOFT;
+    spi_handle->Init.BaudRatePrescaler = SPI_BAUDRATEPRESCALER_16;
+    spi_handle->Init.FirstBit = SPI_FIRSTBIT_MSB;
+    spi_handle->Init.TIMode = SPI_TIMODE_DISABLE;
+    spi_handle->Init.CRCCalculation = SPI_CRCCALCULATION_DISABLE;
+    spi_handle->Init.IOSwap = SPI_IO_SWAP_DISABLE;
+
+    // re-initialize with proper settings
+    if (HAL_SPI_Init(spi_handle) != HAL_OK)
+        return NULL;
+
+    // allocate new struct
+    ssi_encoder_t *encoder = malloc(sizeof(ssi_encoder_t));
+    
+    if (encoder == NULL)
```

Comment:
This will work, I've also just been using `assert()`'s since we don't really want the program to start if hardware isn't initialized correctly